reasoning_gym/algorithmic/jugs.py

```diff
@@ -0,0 +1,314 @@
+
+
+def generate_jug_solution(jug_capacities: Tuple[int, int, int], target: int) -> List[str]:
+    """Solves the jug puzzle and returns a sequence of formatted steps."""
+    capacities = list(jug_capacities)
+    initial_state = (0, 0, 0)
+    queue = deque([(initial_state, [])])
+    visited = set()
+
+    while queue:
+        (state, path) = queue.popleft()
+
+        if target in state:
+            return path  # Solution found
+
+        if state in visited:
+            continue
+        visited.add(state)
+
+        for i in range(3):  # Iterate over each jug
+            # Fill jug i
+            new_state = list(state)
+            new_state[i] = capacities[i]
+            queue.append((tuple(new_state), path + [f"fill {chr(65 + i)}"]))
+
+            # Empty jug i
+            new_state = list(state)
+            new_state[i] = 0
+            queue.append((tuple(new_state), path + [f"empty {chr(65 + i)}"]))
+
+            # Pour from jug i to jug j
+            for j in range(3):
+                if i != j:
+                    new_state = list(state)
+                    pour_amount = min(state[i], capacities[j] - state[j])
+                    new_state[i] -= pour_amount
+                    new_state[j] += pour_amount
+                    queue.append((tuple(new_state), path + [f"pour {chr(65 + i)}->{chr(65 + j)}"]))
+
+    return ["No solution"]  # No valid solution found
+
+
+@dataclass
+class JugsConfig:
+    """Configuration for Jugs puzzle generation"""
+
+    num_jugs: int = 3
+    difficulty: int = 10
+    seed: Optional[int] = None
+    size: int = 500
+
+    def validate(self):
+        """Validate configuration parameters"""
+        assert self.num_jugs > 2, "num_jugs must be gt 2"
+        assert self.difficulty > 0, "difficulty must be gt 0"
+        assert self.difficulty < 200, "difficulty must be lt 200"
+
+
+class JugsDataset(ProceduralDataset):
+    """Generates water jug puzzles inspired by [this scene from _Die Hard 3_](https://www.youtube.com/watch?v=6cAbgAaEOVE), with configurable parameters"""
+
+    def __init__(self, config: JugsConfig):
+        super().__init__(config=config, seed=config.seed, size=config.size)
+
+    def __getitem__(self, idx: int) -> dict:
+        """Generate a single Jugs task
+
+        Returns:
+            dict with keys:
+                - question: str, the task description
+                - answer: str, a solution string
+                - metadata: dict with generation parameters
+        """
+        rng = Random(self.seed + idx)
+
+        puzzle = generate_puzzle(rng, num_jugs=self.config.num_jugs, difficulty=self.config.difficulty)
+        solution = generate_jug_solution(puzzle["jug_capacities"], puzzle["target"])
+
+        cap_str = ", ".join(f"{chr(ord('A')+i)}:{cap}" for i, cap in enumerate(puzzle["jug_capacities"]))
+        question = f"""
+You are a police officer. A maniac has planted a bomb next to a public fountain.
```

Comment:
I woudl remove the newline at the beginnign of this string, e.g. start on the `question = f"""` line.